pkg/ext/store.go

```diff
@@ -2,92 +2,260 @@ package ext
-// meant to control the behavior of the stores. Note: We currently don't have
-// field-manager enabled.
-type Store[T runtime.Object, TList runtime.Object] interface {
-	// Create should store the resource to some backing storage.
-	//
-	// It can apply modifications as necessary before storing it. It must
-	// return a resource of the type of the store, but can
-	// create/update/delete arbitrary objects in Kubernetes without
-	// returning them to the user.
-	//
-	// It is called either when a request creates a resource, or when a
-	// request updates a resource that doesn't exist.
-	Create(ctx Context, obj T, opts *metav1.CreateOptions) (T, error)
-	// Update should overwrite a resource that is present in the backing storage.
-	//
-	// It can apply modifications as necessary before storing it. It must
-	// return a resource of the type of the store, but can
-	// create/update/delete arbitrary objects in Kubernetes without
-	// returning them to the user.
-	//
-	// It is called when a request updates a resource (eg: through a patch or update request)
-	Update(ctx Context, obj T, opts *metav1.UpdateOptions) (T, error)
-	// Get retrieves the resource with the given name from the backing storage.
-	//
-	// Get is called for the following requests:
-	// - get requests: The object must be returned.
-	// - update requests: The object is needed to apply a JSON patch and to make some validation on the change.
-	// - delete requests: The object is needed to make some validation on it.
-	Get(ctx Context, name string, opts *metav1.GetOptions) (T, error)
-	// List retrieves all resources matching the given ListOptions from the backing storage.
-	List(ctx Context, opts *metav1.ListOptions) (TList, error)
-	// Watch sends change events to a returned channel.
-	//
-	// The store is responsible for closing the channel.
-	Watch(ctx Context, opts *metav1.ListOptions) (<-chan WatchEvent[T], error)
-	// Delete deletes the resource of the given name from the backing storage.
-	Delete(ctx Context, name string, opts *metav1.DeleteOptions) error
+// updateValidation is used to do some validation on the object before updating it in the store.
+// One example is running mutating/validating webhooks, though we're not using these yet.
+func CreateOrUpdate[T runtime.Object](
+	ctx context.Context,
+	name string,
+	objInfo rest.UpdatedObjectInfo,
+	createValidation rest.ValidateObjectFunc,
+	updateValidation rest.ValidateObjectUpdateFunc,
+	forceAllowCreate bool,
+	options *metav1.UpdateOptions,
+	getFn func(ctx context.Context, name string, opts *metav1.GetOptions) (T, error),
+	createFn func(ctx context.Context, obj T, opts *metav1.CreateOptions) (T, error),
+	updateFn func(ctx context.Context, obj T, opts *metav1.UpdateOptions) (T, error),
+) (runtime.Object, bool, error) {
+	oldObj, err := getFn(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, false, err
```

Comment:
Here and in all other error returns here - should we wrap and annotate errors? E.g.
```suggestion
			return nil, false, fmt.Errorf("error getting object %s: name, %w", )
```

Reply:
Any errors returned by the various store operations (`Create`, `Update`, `Delete`, `List`, etc) **must** implement [apierrors.APIStatus](https://github.com/kubernetes/apimachinery/blob/v0.31.1/pkg/api/errors/errors.go#L39-L43). If we don't then the library that's calling our store's methods will log an error because it won't recognize the error and the clients (kubectl) won't get the pretty printed error either.

If we just wrap like suggested here, we'd lose the apierrors.APIStatus that `getFn` returns, or createFn / updateFn returns. I could do a better wrapping though if getFn returned the expected APIStatus. Then we can add context to the error message without changing the APIStatus error.

Thoughts?

Reply:
> Any errors returned by the various store operations (Create, Update, Delete, List, etc) must implement apierrors.APIStatus.

Does `create|updateValidation` **always** return `apierrors.APIStatus`?
Also [L152-L153](https://github.com/tomleb/rancher-steve/blob/249d3b8697f98dc6b38d2353357d69d309642e31/pkg/ext/store.go#L152-L153) doesn't conform with the **must** part.

> I could do a better wrapping though if getFn returned the expected APIStatus. Then we can add context to the error message without changing the APIStatus error.

👍

Reply:
> Does create|updateValidation always return apierrors.APIStatus?

I don't think so. We can add our own validation on top of the default provided so it's probably best to to the conversion as a fallback. Added that.

> Also [L152-L153](https://github.com/tomleb/rancher-steve/blob/249d3b8697f98dc6b38d2353357d69d309642e31/pkg/ext/store.go#L152-L153) doesn't conform with the must part.

Yeah, that should be fixed now. I'm doing convertError everywhere except for the getFn / createFn / updateFn calls. I expect them to return `apierrors.APIStatus` errors.